### PR TITLE
Block level bloom filter

### PR DESCRIPTION
**Pull request by GitMark0**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala [hunk 1]
```diff
@@ -133,24 +134,24 @@ class AccountForgeMessageBuilder(
...
+      sidechainTransactions: Seq[SidechainTypes#SCAT],
+      mainchainHeaders: Seq[MainchainHeader],
+      ommers: Seq[Ommer[AccountBlockHeader]],
+      ownerPrivateKey: PrivateKey25519,
+      forgingStakeInfo: ForgingStakeInfo,
+      vrfProof: VrfProof,
+      forgingStakeInfoMerklePath: MerklePath,
+      companion: DynamicTypedSerializer[SidechainTypes#SCAT, TransactionSerializer[SidechainTypes#SCAT]],
+      inputBlockSize: Int,
+      logsBloom: LogsBloom,
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala:152]: we don't need to pass `logsBloom` as an argument here, because:
1) we don't know the value yet
2) it is applicable to the account model only.

Instead inside this method in this class we need to calculate a valid bloom filter using the real set of receipts (see `receiptList`)

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala (hunk 1 above)
Comment (GitMark0) [reply sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala:152]: Good, so it needs to be calculated in 2 places? Here and in `AccountState` (it is already calculated there)-

Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/account/forger/AccountForgeMessageBuilder.scala:152]: right

File: sdk/src/main/scala/com/horizen/account/receipt/LogsBloom.scala [hunk 2]
```diff
@@ -20,25 +25,24 @@ class LogsBloom() {
...
     bloomFilterIndexes.zipWithIndex
       .foreach({ case (bloomFilterIndex, i) =>
-        bloomFilter(bloomFilterIndex) =
-          (bloomFilter(bloomFilterIndex) | bloomFilterValues(i)).toByte
+        bloomFilter(bloomFilterIndex) = (bloomFilter(bloomFilterIndex) | bloomFilterValues(i)).toByte
       })
   }
 
-  def addBloomFilter(bloomFilter: LogsBloom): Unit = {
+  def addBloomFilter(bloomFilter: LogsBloom): LogsBloom = {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/account/receipt/LogsBloom.scala:37]: Let's keep the result type as Unit.
Having `LogsBloom` lead to the idea that addBloomFilter remain the existing instance immutable - which is not our case.
In the forger reuse `fromEthereumReceipt` that should take `EthereumConsensusDataReceipt` seq

Comment (GitMark0) [reply sdk/src/main/scala/com/horizen/account/receipt/LogsBloom.scala:37]: Ok, I added that because it allowed me to use reduce on `EthereumReceipt` seq that was created by `map` from `EthereumConsensusDataReceipt` seq. Should I add another method called `fromEthereumConsensusDataReceipt` to the `LogsBloom` or I should extract `LogsBloom` seq from `EthereumConsensusDataReceipt` to be able to use `fromEthereumReceipt`. Also, approach would be to initialize a new empty `LogsBloom` and add receipts (by calling `addBloomFilter`), one by one to it.

Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/account/receipt/LogsBloom.scala:37]: "or I should extract LogsBloom seq from EthereumConsensusDataReceipt to be able to use fromEthereumReceipt. After that, approach would be to initialize a new empty LogsBloom and add receipts (by calling addBloomFilter), one by one to it." I vote for this one

File: qa/sc_evm_test_block_bloom_filter.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+                                             gasLimit=estimated_gas)
+    print("Generating next block...")
+    generate_next_blocks(node, "first node", 1)
+    tx_receipt = node.rpc_eth_getTransactionReceipt(tx_hash)
+    assert_equal(format_evm(tx_receipt['result']['contractAddress']), format_evm(address))
+    print("Smart contract deployed successfully to address 0x{}".format(address))
+    return address
+
+
+class SCEvmDebugMethods(SidechainTestFramework):
```
Comment (i-Alex) [inline qa/sc_evm_test_block_bloom_filter.py:227]: check the name, seems irrelevant

File: qa/sc_evm_test_block_bloom_filter.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+Test:
+    For the smart contract:
+        - Deploy the smart contract without initial data
+        - Check initial minting success (static_call)
+        - Check initial supply (static_call)
+        - Check a successful transfer (static_call + actual tx)
+        - Check a reverting transfer (static_call + actual tx)
+        - Check approval + transferFrom (static_call + actual tx)
+"""
```
Comment (i-Alex) [inline qa/sc_evm_test_block_bloom_filter.py:39]: update test description

**Review by i-Alex — CHANGES_REQUESTED** [review]
Minor changes in the py test requeired.
Beside that - seems good

File: qa/sc_evm_test_block_bloom_filter.py [hunk 5]
```diff
@@ -224,7 +217,7 @@ def deploy_erc20_smart_contract(node, smart_contract, from_address):
     return address
 
 
-class SCEvmDebugMethods(SidechainTestFramework):
+class SCEvmBlockBloomFilter(SidechainTestFramework):
```
Comment (i-Alex) [inline qa/sc_evm_test_block_bloom_filter.py:220]: check Main function
